src/dune_rules/lib.ml

```diff
@@ -1435,14 +1435,26 @@ end = struct
             in
             (acc_res, acc_selects, acc_re_exports)
           | Direct (loc, name) ->
-            let+ lib = resolve_dep db (loc, name) ~private_deps in
-            let acc_res =
-              let open Resolve.O in
-              let+ lib = lib
-              and+ acc_res = acc_res in
-              lib :: acc_res
-            in
-            (acc_res, acc_selects, acc_re_exports)
+            let ocaml_version = db.lib_config.ocaml_version in
```

Comment:
Not an expert with lib rules, but this feels very early in the pipeline to filter out `bigarray`. For example it will ignore `bigarray` if it's under a `(select)`. I think that it should be done later, after everything has been merged together.

Reply:
This part of the source code is executed Lazily. Almost everything in dune is merged together lazily. Waiting later is like waiting when the package will be read (in dune system resolving the package) to be put in META files like I said https://github.com/ocaml/dune/pull/5526#issue-1175246595, at the reading an error will be raised. Avoiding the error in this function in order to skip generating the package in META files, implies introducing a specific no-op result type and I think that can affect some part of the source code which is not necessary.

I did some inspection, `(select)`  is not used with bigarray right now in opam repository, but `(libraries)` yes:
- `opam grep --regex="(libraries.*bigarray" | wc -l` give 371 package
- `opam grep --regex="(select.*bigarray" | wc -l` give 0 package

Reply:
I think this is the right place, but the other cases need doing. The `Re_export` case is the same as the `Direct` case, so I expect the code here can be hoisted and shared a little (i.e. `(re_export bigarray)` should be a no-op in the same way as `bigarray`).

`select` should be implemented - `!bigarray` should cause a select filter _to fail_ (because bigarray is always there) so a scan of `forbidden` is needed to check for it. For the positive case, `bigarray` should be filtered out of `required`. I quickly tested on 5.0 only with:

```diff
--- a/src/dune_rules/lib.ml
+++ b/src/dune_rules/lib.ml
@@ -1388,11 +1388,15 @@ end = struct
           Memo.Build.List.find_map choices
             ~f:(fun { required; forbidden; file } ->
               let forbidden = Lib_name.Set.to_list forbidden in
+              if List.exists ~f:(fun lib -> String.equal (Lib_name.to_string lib) "bigarray") forbidden
+              then Memo.Build.return None
+              else
               let* exists =
                 Memo.Build.List.exists forbidden ~f:(available_internal db)
               in
               if exists then Memo.Build.return None
               else
+              let required = Lib_name.Set.filter ~f:(fun lib -> not(String.equal (Lib_name.to_string lib) "bigarray")) required in
                 Resolve.Build.peek
                   (let deps =
                      Lib_name.Set.fold required ~init:[] ~f:(fun x acc ->
```
but better code will be available, and it needs the version check. However, those two additions cause this `dune` file to be evaluated correctly in 5.0:

```
(executable
  (name c)
  (libraries (select c.ml from (!bigarray -> c.nobigarray.ml) (bigarray -> c.bigarray.ml) ( -> c.dummy.ml))))
```

without the check on `forbidden`, `c.nobigarray.ml` would be selected and obviously without the filter on `required` the `c.bigarray.ml` would never be selected.

Reply:
Great. Can you include this as a new test case?

Reply:
@emillon regarding the new test case, I'm not familiar with the testsuites(not look in to it before).